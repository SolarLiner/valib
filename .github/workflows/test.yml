name: Tests

on:
  push:
    branches:
      - main
    tags:
      # Run when pushing version tags, since otherwise it's impossible to
      # restart a successful build after pushing a tag
      - '*.*.*'
  pull_request:
    branches:
      - main

defaults:
  run:
    # This otherwise gets run under dash which does not support brace expansion
    shell: bash

jobs:
  test:
    strategy:
      matrix:
        os: [ubuntu-22.04, macos-11, windows-latest]
    name: Build and test all components
    runs-on: ${{ matrix.os }}
    steps:
      - uses: actions/checkout@v2
      # Needed for git-describe to do anything useful
      - name: Fetch all git history
        run: git fetch --force --prune --tags --unshallow

      - name: Install dependencies
        if: startsWith(matrix.os, 'ubuntu')
        run: |
          sudo apt-get update
          sudo apt-get install -y libasound2-dev libgl-dev libjack-dev libx11-xcb-dev  libxcb1-dev libxcb-dri2-0-dev libxcb-icccm4-dev libxcursor-dev libxkbcommon-dev libxcb-shape0-dev libxcb-xfixes0-dev
      - uses: actions/cache@v2
        # FIXME: Caching `target/` causes the Windows runner to blow up after some time
        if: startsWith(matrix.os, 'windows')
        with:
          path: |
            ~/.cargo/registry/index/
            ~/.cargo/registry/cache/
            ~/.cargo/git/db/
          key: ${{ matrix.name }}-${{ matrix.cross-target }}
      - uses: actions/cache@v2
        if: "!startsWith(matrix.os, 'windows')"
        with:
          path: |
            ~/.cargo/registry/index/
            ~/.cargo/registry/cache/
            ~/.cargo/git/db/
            target/
          key: ${{ matrix.name }}-${{ matrix.cross-target }}

      - name: Set up Rust toolchain
        uses: actions-rs/toolchain@v1
      - name: Build all targets
        uses: actions-rs/cargo@v1
        with:
          command: build
          args: --workspace
      - name: Run the tests
        uses: actions-rs/cargo@v1
        with:
          command: test
          args: --workspace
