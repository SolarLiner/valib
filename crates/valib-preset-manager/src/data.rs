use futures::TryStreamExt;
use serde::de::{DeserializeOwned, IntoDeserializer};
use serde::{Deserialize, Serialize};
use smol::io::{AsyncReadExt, AsyncWriteExt};
use smol::stream::{Stream, StreamExt};
use smol::{fs, io, stream};
use std::collections::{BTreeMap, BTreeSet};
use std::path::{Path, PathBuf};
use thiserror::Error;

pub const EXTENSION: &'static str = "preset";

#[allow(unused_variables)]
pub trait PresetData: DeserializeOwned + Serialize {
    const CURRENT_REVISION: u64;

    type PreviousRevision: PresetData;

    fn upgrade(previous: &PresetV1<Self::PreviousRevision>) -> Option<PresetV1<Self>> {
        None
    }
}

impl PresetData for () {
    const CURRENT_REVISION: u64 = 0;
    type PreviousRevision = ();

    fn upgrade(previous: &PresetV1<Self>) -> Option<PresetV1<Self>> {
        Some(previous.clone())
    }
}

impl PresetData for toml::Value {
    const CURRENT_REVISION: u64 = 0;
    type PreviousRevision = toml::Value;

    fn upgrade(previous: &PresetV1<Self::PreviousRevision>) -> Option<PresetV1<Self>> {
        Some(previous.clone())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PresetMeta {
    pub title: String,
    pub author: String,
    pub tags: BTreeSet<String>,
    pub revision: u64,
    #[serde(flatten)]
    pub other: BTreeMap<String, toml::Value>,
}

#[derive(Debug, Copy, Clone, Serialize, Deserialize)]
#[serde(try_from = "u64")]
struct PresetV1Version;

impl TryFrom<u64> for PresetV1Version {
    type Error = String;

    fn try_from(value: u64) -> Result<Self, Self::Error> {
        if value == 1 {
            Ok(Self)
        } else {
            Err(format!(
                "Presets from PresetV1 can only have version = 1 (not {value})"
            ))
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PresetV1<Data> {
    version: PresetV1Version,
    pub metadata: PresetMeta,
    pub data: Data,
}

impl<Data> PresetV1<Data> {
    pub fn new(metadata: PresetMeta, data: Data) -> Self {
        Self {
            version: PresetV1Version,
            metadata,
            data,
        }
    }
}

#[derive(Debug, Error)]
pub enum PresetDeserializeError {
    #[error("Preset I/O Error: {0}")]
    IoError(#[from] io::Error),
    #[error("TOML Parse Error: {0}")]
    TomlError(#[from] toml::de::Error),
}

#[derive(Debug, Error)]
pub enum PresetSerializeError {
    #[error("Preset I/O Error: {0}")]
    IoError(#[from] io::Error),
    #[error("TOML Parse Error: {0}")]
    TomlError(#[from] toml::ser::Error),
}

impl<Data: PresetData> PresetV1<Data> {
    pub async fn from_reader(
        mut reader: impl io::AsyncBufRead + Unpin,
    ) -> Result<Self, PresetDeserializeError> {
        let data = {
            let mut s = String::new();
            reader.read_to_string(&mut s).await?;
            s
        };
        Ok(toml::from_str(&data)?)
    }

    pub async fn from_file(path: &Path) -> Result<Self, PresetDeserializeError> {
        let file = fs::File::open(path).await?;
        let file = io::BufReader::new(file);
        Self::from_reader(file).await
    }

    pub async fn from_title(base_dir: &Path, title: &str) -> Result<Self, PresetDeserializeError> {
        Self::from_file(&base_dir.join(Self::get_preset_filename(title))).await
    }

    pub async fn write_to(
        &self,
        mut writer: impl io::AsyncWrite + Unpin,
        pretty: bool,
    ) -> Result<(), PresetSerializeError> {
        let data = if pretty {
            toml::to_string_pretty(self)
        } else {
            toml::to_string(self)
        }?;
        writer.write_all(data.as_bytes()).await?;
        Ok(())
    }

    pub async fn write_to_file(
        &self,
        path: &Path,
        pretty: bool,
    ) -> Result<(), PresetSerializeError> {
        let file = fs::File::create(path).await?;
        let file = io::BufWriter::new(file);
        self.write_to(file, pretty).await
    }

    pub async fn write_to_bank(
        &self,
        base_dir: &Path,
        pretty: bool,
    ) -> Result<(), PresetSerializeError> {
        self.write_to_file(&base_dir.join(self.preset_filename()), pretty)
            .await
    }

    pub fn get_preset_filename(title: &str) -> PathBuf {
        Path::new(title).with_extension(EXTENSION)
    }

    pub fn preset_filename(&self) -> PathBuf {
        Self::get_preset_filename(&self.metadata.title)
    }
}

pub fn read_folder<Data: PresetData>(
    dir: &Path,
) -> impl Stream<Item = Result<PresetV1<Data>, PresetDeserializeError>> + '_ {
    stream::once_future(fs::read_dir(dir))
        .try_flatten_unordered(None)
        .try_filter_map(|entry| async move {
            let metadata = entry.metadata().await?;
            if metadata.is_file() {
                Ok(Some(entry.path()))
            } else {
                Ok(None)
            }
        })
        .map(|res| res.map_err(PresetDeserializeError::IoError))
        .try_filter_map(|path| async move {
            let preset = PresetV1::from_file(&path).await?;
            Ok(Some(preset))
        })
}
