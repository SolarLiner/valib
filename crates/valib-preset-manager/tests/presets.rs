use rstest::rstest;
use smol_macros::test;
use std::path::PathBuf;
use valib_preset_manager::data::PresetV1;

#[rstest]
fn read_presets_good(#[files("tests/banks/Good/*.preset")] preset_path: PathBuf) {
    let preset = smol::block_on(PresetV1::<toml::Value>::from_file(&preset_path))
        .expect("Preset should load correctly");
    insta::assert_debug_snapshot!(
        preset_path
            .file_stem()
            .unwrap()
            .to_string_lossy()
            .to_string(),
        preset
    );
}

#[rstest]
fn read_presets_bad(#[files("tests/banks/Bad/*.preset")] preset_path: PathBuf) {
    let preset = smol::block_on(PresetV1::<toml::Value>::from_file(&preset_path))
        .expect_err("Preset should not load correctly");
    insta::assert_debug_snapshot!(
        preset_path
            .file_stem()
            .unwrap()
            .to_string_lossy()
            .to_string(),
        preset
    );
}
