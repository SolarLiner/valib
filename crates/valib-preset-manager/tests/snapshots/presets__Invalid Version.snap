---
source: crates/valib-preset-manager/tests/presets.rs
expression: preset
---
TomlError(
    Error {
        inner: Error {
            inner: TomlError {
                message: "Presets from PresetV1 can only have version = 1 (not 2)",
                raw: Some(
                    "version = 2\r\n\r\n[metadata]\r\ntitle = \"Invalid Version\"\r\nauthor = \"valib\"\r\nrevision = 1\r\ntags = [\"default\",\"init\"]\r\n\r\n[data]",
                ),
                keys: [
                    "version",
                ],
                span: Some(
                    10..11,
                ),
            },
        },
    },
)
